pkg/nginxprocess/process.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+
+// Package nginxprocess contains utilities for working with OS-level NGINX processes.
+package nginxprocess
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/shirou/gopsutil/v4/process"
+)
+
+// Process contains a snapshot of read-only data about an OS-level NGINX process. Create using [List] or [Find].
+type Process struct {
+	Created time.Time
+	Name    string
+	Cmd     string
+	Exe     string // path to the executable
+	Status  string // process status, only present if this process was created using [WithStatus]
+	PID     int32
+	PPID    int32 // parent PID
+}
+
+// IsWorker returns true if the process is a NGINX worker process.
+func (p *Process) IsWorker() bool { return strings.Contains(p.Cmd, "worker") }
+
+// IsShuttingDown returns true if the process is shutting down. This can identify workers that are in the process of a
+// graceful shutdown. See [changing NGINX configuration] for more details.
+//
+// [changing NGINX configuration]: https://nginx.org/en/docs/control.html#reconfiguration
+func (p *Process) IsShuttingDown() bool { return strings.Contains(p.Cmd, "is shutting down") }
+
+// IsHealthy uses Status flags to judge process health. Only works on processes created using [WithStatus].
+func (p *Process) IsHealthy() bool {
+	return p.Status != "" && !strings.Contains(p.Status, process.Zombie)
+}
+
+type options struct {
+	loadStatus bool
+}
+
+// Option customizes how processes are gathered from the OS.
+type Option interface{ apply(opts *options) }
+
+type optionFunc func(*options)
+
+func (f optionFunc) apply(o *options) { f(o) }
+
+// WithStatus runs an additional lookup to load the process status.
+func WithStatus(v bool) Option { //nolint:ireturn // functional options can be opaque
+	return optionFunc(func(o *options) { o.loadStatus = v })
+}
+
+func convert(ctx context.Context, p *process.Process, o options) (*Process, error) {
+	if ctx.Err() != nil { // fail fast if we've canceled
+		return nil, ctx.Err()
+	}
+
+	name, _ := p.NameWithContext(ctx) // slow: shells out to ps
+	if name != "nginx" {
+		return nil, errNotAnNginxProcess
+	}
+
+	cmdLine, _ := p.CmdlineWithContext(ctx) // slow: shells out to ps
+	// ignore nginx processes in the middle of an upgrade
+	if !strings.HasPrefix(cmdLine, "nginx:") || strings.Contains(cmdLine, "upgrade") {
+		return nil, errNotAnNginxProcess
+	}
+
+	var status string
+	if o.loadStatus {
+		flags, _ := p.StatusWithContext(ctx) // slow: shells out to ps
+		status = strings.Join(flags, " ")
+	}
+
+	// unconditionally run fast lookups
+	var created time.Time
+	if millisSinceEpoch, err := p.CreateTimeWithContext(ctx); err == nil {
+		nanos := millisSinceEpoch * int64(time.Millisecond)
+		created = time.Unix(0, nanos)
```

Comment:
confused, why not `time. UnixMilli(millisSinceEpoch)`, https://pkg.go.dev/time#UnixMilli

Reply:
> why not ...

because I didn't know `UnixMilli` existed 😄  

done